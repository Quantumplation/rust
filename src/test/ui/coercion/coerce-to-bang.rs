#![feature(never_type)]

fn foo(x: usize, y: !, z: usize) { }

fn call_foo_a() {
    foo(return, 22, 44);
    //~^ ERROR arguments to this function are incorrect
}

fn call_foo_b() {
    // Divergence happens in the argument itself, definitely ok.
    foo(22, return, 44);
}

fn call_foo_c() {
    // This test fails because the divergence happens **after** the
    // coercion to `!`:
    foo(22, 44, return); //~ ERROR arguments to this function are incorrect
}

fn call_foo_d() {
    // This test passes because `a` has type `!`:
    let a: ! = return;
    let b = 22;
    let c = 44;
    foo(a, b, c); // ... and hence a reference to `a` is expected to diverge.
    //~^ ERROR arguments to this function are incorrect
}

fn call_foo_e() {
    // This test probably could pass but we don't *know* that `a`
    // has type `!` so we don't let it work.
    let a = return;
    let b = 22;
    let c = 44;
    foo(a, b, c); //~ ERROR arguments to this function are incorrect
}

fn call_foo_f() {
    // This fn fails because `a` has type `usize`, and hence a
    // reference to is it **not** considered to diverge.
    let a: usize = return;
    let b = 22;
    let c = 44;
    foo(a, b, c); //~ ERROR arguments to this function are incorrect
}

fn array_a() {
    // Return is coerced to `!` just fine, but `22` cannot be.
    let x: [!; 2] = [return, 22]; //~ ERROR arguments to this function are incorrect
}

fn array_b() {
    // Error: divergence has not yet occurred.
    let x: [!; 2] = [22, return]; //~ ERROR arguments to this function are incorrect
}

fn tuple_a() {
    // No divergence at all.
    let x: (usize, !, usize) = (22, 44, 66);
    //~^ ERROR arguments to this function are incorrect
}

fn tuple_b() {
    // Divergence happens before coercion: OK
    let x: (usize, !, usize) = (return, 44, 66);
    //~^ ERROR arguments to this function are incorrect
}

fn tuple_c() {
    // Divergence happens before coercion: OK
    let x: (usize, !, usize) = (22, return, 66);
}

fn tuple_d() {
    // Error: divergence happens too late
    let x: (usize, !, usize) = (22, 44, return);
    //~^ ERROR arguments to this function are incorrect
}

fn main() { }
