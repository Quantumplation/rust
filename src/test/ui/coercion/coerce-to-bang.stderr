error[E0308]: arguments to this function are incorrect
  --> $DIR/coerce-to-bang.rs:6:5
   |
LL |     foo(return, 22, 44);
   |     ^^^^------^^^^^^^^^
   |         |
   |         expected `!`, found `!`
   |
note: function defined here
  --> $DIR/coerce-to-bang.rs:3:4
   |
LL | fn foo(x: usize, y: !, z: usize) { }
   |    ^^^ --------  ----  --------
help: provide an argument of the correct type
   |
LL |     foo(return, {!}, 44);
   |     ^^^^^^^^^^^^^^^^^^^^

error[E0308]: arguments to this function are incorrect
  --> $DIR/coerce-to-bang.rs:18:5
   |
LL |     foo(22, 44, return);
   |     ^^^^--^^^^^^^^^^^^^
   |         |
   |         expected `!`, found `usize`
   |
note: function defined here
  --> $DIR/coerce-to-bang.rs:3:4
   |
LL | fn foo(x: usize, y: !, z: usize) { }
   |    ^^^ --------  ----  --------
help: provide an argument of the correct type
   |
LL |     foo(22, {!}, return);
   |     ^^^^^^^^^^^^^^^^^^^^

error[E0308]: arguments to this function are incorrect
  --> $DIR/coerce-to-bang.rs:26:5
   |
LL |     foo(a, b, c); // ... and hence a reference to `a` is expected to diverge.
   |     ^^^^-^^^^^^^
   |         |
   |         expected `!`, found `!`
   |
note: function defined here
  --> $DIR/coerce-to-bang.rs:3:4
   |
LL | fn foo(x: usize, y: !, z: usize) { }
   |    ^^^ --------  ----  --------
help: provide an argument of the correct type
   |
LL |     foo(a, {!}, c); // ... and hence a reference to `a` is expected to diverge.
   |     ^^^^^^^^^^^^^^

error[E0308]: arguments to this function are incorrect
  --> $DIR/coerce-to-bang.rs:36:5
   |
LL |     foo(a, b, c);
   |     ^^^^-^^^^^^^
   |         |
   |         expected `!`, found `_`
   |
note: function defined here
  --> $DIR/coerce-to-bang.rs:3:4
   |
LL | fn foo(x: usize, y: !, z: usize) { }
   |    ^^^ --------  ----  --------
help: provide an argument of the correct type
   |
LL |     foo(a, {!}, c);
   |     ^^^^^^^^^^^^^^

error[E0308]: arguments to this function are incorrect
  --> $DIR/coerce-to-bang.rs:45:5
   |
LL |     foo(a, b, c);
   |     ^^^^-^^^^^^^
   |         |
   |         expected `!`, found `usize`
   |
note: function defined here
  --> $DIR/coerce-to-bang.rs:3:4
   |
LL | fn foo(x: usize, y: !, z: usize) { }
   |    ^^^ --------  ----  --------
help: provide an argument of the correct type
   |
LL |     foo(a, {!}, c);
   |     ^^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> $DIR/coerce-to-bang.rs:50:21
   |
LL |     let x: [!; 2] = [return, 22];
   |            ------   ^^^^^^^^^^^^ expected `!`, found integer
   |            |
   |            expected due to this
   |
   = note: expected array `[!; 2]`
              found array `[{integer}; 2]`

error[E0308]: mismatched types
  --> $DIR/coerce-to-bang.rs:55:22
   |
LL |     let x: [!; 2] = [22, return];
   |                      ^^ expected `!`, found integer
   |
   = note: expected type `!`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/coerce-to-bang.rs:60:37
   |
LL |     let x: (usize, !, usize) = (22, 44, 66);
   |                                     ^^ expected `!`, found integer
   |
   = note: expected type `!`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/coerce-to-bang.rs:66:41
   |
LL |     let x: (usize, !, usize) = (return, 44, 66);
   |                                         ^^ expected `!`, found integer
   |
   = note: expected type `!`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/coerce-to-bang.rs:77:37
   |
LL |     let x: (usize, !, usize) = (22, 44, return);
   |                                     ^^ expected `!`, found integer
   |
   = note: expected type `!`
              found type `{integer}`

error: aborting due to 10 previous errors

For more information about this error, try `rustc --explain E0308`.
